### PR TITLE
TEST-#2509: Io tests refactoring

### LINKED ISSUE #2509: IO tests refactoring
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. What kind of performance improvements would you like to see with this new API?
Entire IO test (`modin/pandas/test/test_io.py`) should be refactored in order to combine similar tests in classes. Additionally tests should be reworked to provide capability to run all IO tests in parallel (parallel execution is needed for running tests in parallel using TeamCity with increased size dataset).

### PR DESCRIPTION
**Pull request by amyskov**
## What do these changes do?
This PR refactors IO tests in order to unify source code, group similar tests in classes and capability to run io tests in parallel. Changes are next:
1) Similar tests were grouped in classes.
2) Source code unifications - used common `eval_io`, `eval_from_str`, ... functions, unique test filenames for each tests are used (needed to provide capability to run io tests in parallel), some duplicated code removement.
3) Add capability to process warnings by `eval_general` function.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/CONTRIBUTING.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2509  
- [x] tests added and passing

### CODE REVIEW

File: modin/pandas/test/test_io.py [hunk 1]
```diff
@@ -55,20 +55,18 @@
...
-TEST_JSON_FILENAME = "test.json"
-TEST_HTML_FILENAME = "test.html"
-TEST_EXCEL_FILENAME = "test.xlsx"
-TEST_FEATHER_FILENAME = "test.feather"
-TEST_READ_HDF_FILENAME = "test.hdf"
-TEST_WRITE_HDF_FILENAME_MODIN = "test_write_modin.hdf"
-TEST_WRITE_HDF_FILENAME_PANDAS = "test_write_pandas.hdf"
-TEST_STATA_FILENAME = "test.dta"
-TEST_PICKLE_FILENAME = "test.pkl"
-TEST_SAS_FILENAME = os.getcwd() + "/data/test1.sas7bdat"
```
Comment (anmyachev) [inline modin/pandas/test/test_io.py:69]: Why are you remove this variable and test that using it?

File: modin/pandas/test/test_io.py [hunk 2]
```diff
@@ -93,9 +99,14 @@ def make_parquet_file():
     Yields:
         Function that generates a parquet file/dir
     """
+    filenames = []
 
     def _make_parquet_file(
-        row_size=NROWS, force=False, directory=False, partitioned_columns=[]
+        filename=TEST_PARQUET_FILENAME,
```
Comment (anmyachev) [inline modin/pandas/test/test_io.py:105]: update docstring for `filename` argument

File: modin/pandas/test/test_io.py [hunk 3]
```diff
@@ -175,7 +161,16 @@ def assert_files_eq(path1, path2):
 
 def teardown_test_file(test_path):
     if os.path.exists(test_path):
-        os.remove(test_path)
+        # PermissionError can occure because of issue #2533
```
Comment (anmyachev) [inline modin/pandas/test/test_io.py:164]: Does this occur when files with the same names are created in parallel in different processes?

File: modin/pandas/test/test_io.py [hunk 4]
```diff
@@ -352,135 +342,94 @@ def TestReadCSVFixture():
...
-            try:
-                os.remove(filename)
-            except PermissionError:
-                pass
+    teardown_test_files(filenames)
 
 
-def setup_json_file(row_size, force=False):
-    if os.path.exists(TEST_JSON_FILENAME) and not force:
+def setup_json_file(filename=TEST_JSON_FILENAME, row_size=NROWS, force=True):
```
Comment (anmyachev) [inline modin/pandas/test/test_io.py:348]: Why `force=True`?

File: modin/pandas/test/test_io.py [hunk 5]
```diff
@@ -546,26 +493,31 @@ def setup_fwf_file(overwrite=False, fwf_data=None):
...
         f.write(fwf_data)
 
 
-def teardown_fwf_file():
-    if os.path.exists(TEST_FWF_FILENAME):
-        try:
-            os.remove(TEST_FWF_FILENAME)
-        except PermissionError:
-            pass
+def eval_to_file(modin_obj, pandas_obj, fn, extension, **fn_kwargs):
```
Comment (anmyachev) [inline modin/pandas/test/test_io.py:500]: add docstring

File: modin/pandas/test/test_io.py [hunk 6]
```diff
@@ -1102,1229 +1054,1120 @@ def test_read_csv_internal(
...
+        pandas_df = pandas.read_csv(dataset_url)
 
+        # This first load is to trigger all the import deprecation warnings
+        modin_df = pd.read_csv(dataset_url)
 
-def test_from_parquet_partition_with_columns(make_parquet_file):
-    make_parquet_file(NROWS, directory=True)
+        # This will warn if it defaults to pandas behavior, but it shouldn't
+        with pytest.warns(None) as record:
+            modin_df = pd.read_csv(dataset_url)
```
Comment (anmyachev) [inline modin/pandas/test/test_io.py:1129]: Why do we do this check in this case, but not in the rest.

Do we really need it?

**Review by anmyachev — CHANGES_REQUESTED** [review]

File: modin/pandas/test/test_io.py [hunk 7]
```diff
@@ -1102,1229 +1054,1120 @@ def test_read_csv_internal(
...
+        unique_filename = get_unique_filename(extension=None)
+        make_parquet_file(filename=unique_filename, partitioned_columns=["col1"])
+        eval_io(
+            fn_name="read_parquet",
+            # read_parquet kwargs
+            path=unique_filename,
+        )
 
-    df_equals(modin_df, pandas_df)
+    def test_read_parquet_partitioned_columns_with_columns(self, make_parquet_file):
```
Comment (anmyachev) [inline modin/pandas/test/test_io.py:1362]: It seems to me that `test_read_parquet_partitioned_columns_with_columns` can be easy merged with `test_read_parquet_partitioned_columns`.

**Review by anmyachev — CHANGES_REQUESTED** [review]

File: modin/pandas/test/test_io.py (hunk 1 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py:69]: This variable refers to non-existent file and test, that uses this variable doesn't create this file (also this test marked as `pytest.skip`).

File: modin/pandas/test/test_io.py (hunk 2 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py:105]: Docstring was updated.

File: modin/pandas/test/test_io.py (hunk 3 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py:165]: No, it affects both parallel and sequential execution modes.

File: modin/pandas/test/test_io.py (hunk 4 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py:349]: Potentially, if default `filename` is used, setup_... functions are changed and old data file is still exists, these data files won't be replaced with new ones. With `force=True` we eliminate this possibility.

File: modin/pandas/test/test_io.py (hunk 5 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py:501]: Docstring is added.

File: modin/pandas/test/test_io.py (hunk 6 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py]: I think we can use regular `eval_io` function here.

File: modin/pandas/test/test_io.py (hunk 7 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py]: Ok, tests are merged.

File: modin/pandas/test/utils.py [hunk 8]
```diff
@@ -645,6 +645,7 @@ def eval_general(
     check_exception_type=True,
     raising_exceptions=None,
     check_kwargs_callable=True,
+    modin_warning=False,
```
Comment (anmyachev) [inline modin/pandas/test/utils.py:648]: It's better to use `None` here.

File: modin/pandas/test/test_io.py (hunk 4 above)
Comment (anmyachev) [reply modin/pandas/test/test_io.py:349]: May simply not provide a default value for `filename`?

File: modin/pandas/test/utils.py (hunk 8 above)
Comment (anmyachev) [reply modin/pandas/test/utils.py:648]: Let's leave the functionality for checking warnings outside of `eval_general`. It's already the super-mega-power function.

It's a pretty nice as is.
```python
with pytest.warns(UserWarning):
    eval_general(...)
```

**Review by anmyachev — CHANGES_REQUESTED** [review]

File: modin/pandas/test/utils.py (hunk 8 above)
Comment (amyskov) [reply modin/pandas/test/utils.py]: Ok, both suggestions were implemented.

File: modin/pandas/test/test_io.py (hunk 4 above)
Comment (amyskov) [reply modin/pandas/test/test_io.py]: Ok, default value was removed.

**Review by anmyachev — APPROVED** [review]
LGTM!

**Comment by anmyachev** [comment]
@YarShev can we merge this?

**Comment by YarShev** [comment]
@anmyachev , yes, looks good!

**Review by YarShev — APPROVED** [review]

**Pull request merged**